pkg/kds/mux/session_test.go

```diff
@@ -0,0 +1,115 @@
+			serverSession = mux.NewSession("remote-1", &testMultiplexStream{input: output, output: input}, nil)
+		})
+
+		It("should Send to clientSession's ClientStream and Recv from serverSession's ServerStream", func() {
+			err := clientSession.ClientStream().Send(&envoy_api_v2.DiscoveryRequest{VersionInfo: "1"})
+			Expect(err).ToNot(HaveOccurred())
+			msg, err := serverSession.ServerStream().Recv()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(msg.VersionInfo).To(Equal("1"))
+		})
+		It("should Send to serverSession's ServerStream and Recv from clientSession's ClientStream", func() {
+			err := serverSession.ServerStream().Send(&envoy_api_v2.DiscoveryResponse{VersionInfo: "2"})
+			Expect(err).ToNot(HaveOccurred())
+			msg, err := clientSession.ClientStream().Recv()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(msg.VersionInfo).To(Equal("2"))
+		})
+		It("should Send to clientSession's ServerStream and Recv from serverSession's ClientStream", func() {
+			err := clientSession.ServerStream().Send(&envoy_api_v2.DiscoveryResponse{VersionInfo: "3"})
+			Expect(err).ToNot(HaveOccurred())
+			msg, err := serverSession.ClientStream().Recv()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(msg.VersionInfo).To(Equal("3"))
+		})
+		It("should Send to serverSession's ClientStream and Recv from clientSession's ServerStream", func() {
+			err := serverSession.ClientStream().Send(&envoy_api_v2.DiscoveryRequest{VersionInfo: "4"})
+			Expect(err).ToNot(HaveOccurred())
+			msg, err := clientSession.ServerStream().Recv()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(msg.VersionInfo).To(Equal("4"))
+		})
+	})
+
+	Context("concurrent operations", func() {
+		var clientSession mux.Session
+		var serverSession mux.Session
+
+		BeforeEach(func() {
+			input := make(chan *mesh_proto.Message, 1)
+			output := make(chan *mesh_proto.Message, 1)
+			clientSession = mux.NewSession("global", &testMultiplexStream{input: input, output: output}, nil)
+			serverSession = mux.NewSession("remote-1", &testMultiplexStream{input: output, output: input}, nil)
+		})
+
+		Context("Recv", func() {
+			It("should block while proper Send called", func() {
+				wait := make(chan struct{})
+				go func() {
+					request, err := serverSession.ServerStream().Recv()
+					Expect(err).ToNot(HaveOccurred())
+					Expect(request.VersionInfo).To(Equal("1"))
+					close(wait)
+				}()
+				err := clientSession.ServerStream().Send(&envoy_api_v2.DiscoveryResponse{VersionInfo: "2"})
+				Expect(err).ToNot(HaveOccurred())
+				err = clientSession.ClientStream().Send(&envoy_api_v2.DiscoveryRequest{VersionInfo: "1"})
+				Expect(err).ToNot(HaveOccurred())
+
+				resp, err := serverSession.ClientStream().Recv()
+				Expect(err).ToNot(HaveOccurred())
+				Expect(resp.VersionInfo).To(Equal("2"))
+
+				select {
+				case <-wait:
+				case <-time.After(1 * time.Second):
+					err = errors.New("timeout waiting for ServerStream.Recv()")
+				}
+				Expect(err).ToNot(HaveOccurred())
+			})
```

Comment:
How about this
```
It("should block while proper Send called", func(done Done) {
    go func() {
        request, err := serverSession.ServerStream().Recv()
        Expect(err).ToNot(HaveOccurred())
        Expect(request.VersionInfo).To(Equal("1"))
        close(done)
    }()
    err := clientSession.ServerStream().Send(&envoy_api_v2.DiscoveryResponse{VersionInfo: "2"})
    Expect(err).ToNot(HaveOccurred())
    err = clientSession.ClientStream().Send(&envoy_api_v2.DiscoveryRequest{VersionInfo: "1"})
    Expect(err).ToNot(HaveOccurred())

    resp, err := serverSession.ClientStream().Recv()
    Expect(err).ToNot(HaveOccurred())
    Expect(resp.VersionInfo).To(Equal("2"))
}, "1s")
```

Reply:
Thanks, much better! Didn't know about `done Done`